pkg/cmd/cli/describe/chaindescriber.go

```diff
@@ -189,3 +192,46 @@ func outputHelper(info, namespace string, singleNamespace bool) string {
 	}
 	return fmt.Sprintf("<%s %s>", namespace, info)
 }
+
+// DepthFirst implements stateful depth-first graph traversal.
+// Modifies behavior of visitor.DepthFirst to allow nodes to be visited multiple
+// times as long as they're not in the current stack
+type DepthFirst struct {
+	EdgeFilter func(graph.Edge) bool
+	Visit      func(u, v graph.Node)
+	stack      internal.NodeStack
+}
+
+// Walk performs a depth-first traversal of the graph g starting from the given node
+func (d *DepthFirst) Walk(g graph.Graph, from graph.Node, until func(graph.Node) bool) graph.Node {
+	d.stack.Push(from)
+
+	for d.stack.Len() > 0 {
+		t := d.stack.Pop()
+		if until != nil && until(t) {
+			return t
+		}
+		for _, n := range g.From(t) {
+			if d.EdgeFilter != nil && !d.EdgeFilter(g.Edge(t, n)) {
+				continue
+			}
+			if d.Visited(n.ID()) {
+				continue
+			}
+			if d.Visit != nil {
+				d.Visit(t, n)
+			}
+			d.stack.Push(n)
+		}
+	}
+	return nil
+}
+
+func (d *DepthFirst) Visited(id int) bool {
```

Comment:
If you have a simple loop `1->2` and `2->1`, this won't correctly determine the visited status.  The stack structure makes the depth first traversal "easy", but it also loses the information about the chain that brought you to exploring a particular node.